### ISSUE TITLE
Static analysis of module attributes in twisted.python.modules

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4531]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#4531|
|Type|enhancement|
|Created|2010-06-30 06:32:06Z|
|Branch|https://github.com/twisted/twisted/tree/static-modules-4531-2|

Python 2.6's `ast` module makes it easy to find imports, exports, and defined names in a module. I've added support to `twisted.python.modules` for doing this, as per the docstring describing this as 'future work'.

<details><summary>Searchable metadata</summary>

```
trac-id__4531 4531
type__enhancement enhancement
reporter__washort washort
priority__normal normal
milestone__ 
branch__branches_static_modules_4531_2 branches/static-modules-4531-2
branch_author__therve__washort__glyph therve, washort, glyph
status__new new
resolution__None None
component__core core
keywords__ 
time__1277879526000000 1277879526000000
changetime__1378811726000000 1378811726000000
version__None None
owner__therve therve
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_1"></a>|@washort commented|
|-|-|

(In [29492]) Branching to 'static-modules-4531'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @washort|
|-|-|

Awesome!  This is a very cool feature.  `twisted.python.modules` is super.

  1. twisted/test/test_modules.py
    1. ast module import should go with the other stdlib module imports
    1. Use `assertEquals` instead of `assertEqual`
    1. Whitespace around operators (eg `modinfo = modules.getModule(self.packageName+".a")`)
    1. `test_moduleExportNames` looks like two unit tests, one for the `__all__` case, one for the no-`__all__` case.  Can you split it into two methods?
    1. 
  1. twisted/python/modules.py
    1. Ugh, indented class definition.  I have nightmares about these things from conch.  Can we avoid the conditional definition?  Maybe just select the base class conditionally?  Clearly the resulting class won't work without the ast module, but having to juggle name usage propagates the awkwardness ever outwards.  It's easier to always use the name but sometimes not use the object.
    1. But speaking of outward propagation, `ImportExportFinder` doesn't need to be public right?
    1. There's no Python code coverage tool that will tell me if every case of `visit_Module` is covered.  It'd be easier to be confident of its coverage if there were some direct unit tests for `ImportExportFinder`.  The new tests on `PathModificationTest` are good, but it'll be easier to see all the conditionals in `visit_Module` being treated if there's a unit test specifically targeted at each one, I think.
    1. `PythonModule.finder` should be private, I think.
    1. And documented.
    1. And there's some missing blank lines after it and after `_getEntry`
    1. `PythonAttribute.iterAttributes` needs a docstring
    1. `PythonAttribute`'s methods all lack the proper number of separating blank lines.
    1. In the `PythonModule.iterAttributes` docstring, should `inspect.getmodule` be `inspect.getmemebers`?
    1. In `PythonModule.iterAttributes`, missing whitespace around operator `+` in a few places.
    1. The exception raised by `PythonModule.iterImportNames` looks like it has the wrong message.  This isn't about unloaded modules right?
    1. `PythonModule.iterImportNames` should document that it can raise `NotImplementedError`
    1. Previous two comments also apply to `PythonModule.iterExportNames` I think.
    1. Despite what I said above about the shortcomings of python coverage tools, coverage.py does actually reveal some uncovered code introduced or changed in this branch. :)  Please also add tests to make sure everything is covered.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_4"></a>|@washort removed owner|
|-|-|

I believe I've addressed all these points; the only one I'm not sure about is #11 -- the exception is about how these operations on unloaded modules aren't supported if you don't have `ast`. Also it looks like `./bin/trial --coverage twisted.test.test_modules.PathModificationTest ` shows 100% coverage for `_ImportExportFinder`. Should the separate tests for it remain?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @washort|
|-|-|

> the only one I'm not sure about is #11 -- the exception is about how these operations on unloaded modules aren't supported if you don't have ast.

Specifically for the `NotImplementedError` raised by `iterExportNames`, I find the message confusing because of this possible scenario (using Python 2.5):

```
>>> from twisted.python.modules import getModule
>>> m = getModule('filepath')
>>> m.isLoaded()
True
>>> m.iterExportNames()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "twisted/python/modules.py", line 506, in iterExportNames
    raise NotImplementedError("Examining attributes of unloaded modules"
NotImplementedError: Examining attributes of unloaded modules requires the 'ast' module, found in Python 2.6 or later.
>>> 
```

Another issue I noticed:

```
>>> from twisted.python.modules import getModule
>>> m = getModule('pickle')
>>> m.filePath
FilePath('/usr/lib/python2.6/pickle.pyc')
>>> m.iterExportNames()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/branches/static-modules-4531/twisted/python/modules.py", line 509, in iterExportNames
    self._maybeLoadFinder()
  File "/home/exarkun/Projects/Twisted/branches/static-modules-4531/twisted/python/modules.py", line 438, in _maybeLoadFinder
    self._finder.visit(ast.parse(self.filePath.getContent()))
  File "/usr/lib/python2.6/ast.py", line 37, in parse
    return compile(expr, filename, mode, PyCF_ONLY_AST)
TypeError: compile() expected string without null bytes
>>> 
```

A couple other things that I think are misbehaviors:

```
>>> from twisted.python.modules import getModule
>>> m = getModule('twisted.mail.pop3client')
>>> a = list(m.iterAttributes())[0]
>>> a
PythonAttribute<'twisted.mail.pop3client.LineTooLong'>
>>> a.isLoaded()
False
>>> a.load()                # <----- shouldn't return None
>>> m.load()
<module 'twisted.mail.pop3client' from '/home/exarkun/Projects/Twisted/branches/static-modules-4531/twisted/mail/pop3client.py'>
>>> a = list(m.iterAttributes())[0]
>>> a.isLoaded()
True
>>> a.load()
'-ERR'
>>> 
```


```
>>> from twisted.python.modules import getModule
>>> m = getModule('filepath')
>>> list(m.iterAttributes())
[PythonAttribute<'filepath.__version__'>, PythonAttribute<'filepath.__all__'>, PythonAttribute<'filepath.__version_info__'>]
>>> list(m.iterExportNames())
['InsecurePath', 'LinkError', 'UnlistableError', 'FilePath', 'ZipArchive', 'ZipPath']
>>> 
```

Test coverage of the new code looks better now.  I use coverage.py to get coverage reports now, it's a lot better than trial's built-in coverage stuff.  However, neither can report things like:

  1. In `visit_Module`, the `and len(stmt.targets) == 1` is untested.  It's on an executed line, and it may even be evaluated, but I can delete it without breaking any tests.
  1. The same goes for the next part of the condition, `and isinstance(stmt.targets[0], ast.Name)`
  1. And for `not (isinstance(stmt.value, (ast.List, ast.Tuple)))`.

That might be all of the uncovered parts in `visit_Module`, but it might not hurt to reconstruct this method TDD to be sure.  These are the test cases I think will be much clearer on `ASTVisitorTests` than on `PathModificationTest`.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_6"></a>|@washort removed owner|
|-|-|

Thanks for the review. I've redone the AST code and added new tests to address the points you mentioned.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to @washort|
|-|-|

Minor but looks like ```NotLoadedError``` needs an ```@since``` marker.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort removed owner|
|-|-|

added @since to `NotLoadedError`, `iterExportNames`, and `iterImportNames`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @washort|
|-|-|


I did find a few more things while I was reviewing this, but don't worry: this
isn't nearly as much work as it probably looks like on first glance.

    1. test_lookForSource's docstring should be a bit more expository.  If a
       .py file exists, as opposed to... what?  This is python, surely a .py
       file exists ''somewhere'' :).
    2. Possibly for a separate ticket, but: I should be able to unambiguously
       request the scan-the-source behavior, because sometimes it differs from
       the inspect-the-loaded-module behavior, and I'd like to be able to
       control which one I get.  Some of the tests, like test_attributeLoading,
       use getModule() and then have to perform sanity checks to make sure that
       they're getting the right codepath, ("`assertEquals(....isLoaded(), False)`").  `iterAttributes()` as an API meaning "do your best, and give me the
       most accurate version of the attribute list that you can, but don't load
       any python code that isn't already" is a fine idea, so ''some'' test
       like that should exist, but you should be able to unambiguously invoke
       the other behavior without smashing global state like `sys.modules`.
    3. `iterAttributes` wasn't a really great name. "iter" is one of those
       prefixes like "get" which can be unhelpfully applied to a huge number of
       APIs; basically anything that returns a collection.  python itself is
       moving in this direction (c.f. py3k 'iterkeys', 'itervalues' `=>`
       'keys', 'values')  I'd like the new APIs to be called exportedNames()
       and importedNames() rather than continuing in that flawed pattern.
    4. It ''is'' possible to do something with code that does "`import *`", by
       recursively looking at the exported names of the module that they're
       being imported from.  In fact, that would be really handy.  This can
       definitely wait for a separate ticket, but `test_noImportStar` should
       cite that ticket, not make it sound like an impossibility.
    5. The definition of "exported names" is a bit too loose for my taste.  We
       should implement 'exported' to be what Python implements for `import *`,
       which is that you get all defined names that don't start with "_",
       unless you define `__all__`.
    6. My secret plan for this code is to get Pydoctor to adopt it, and a
       documentation generator should be able to easily compute the list of names
       which should appear to be public, which is: exported names, unless
       `__all__` is undefined, in which case the names you would get from `import
       *` minus imported names.  Having this implemented in
       `twisted.python.modules` itself would be a bonus, because it would allow
       us to use the exact same rules for an import hook or module proxy or
       something that would say "only allow importing of names that the
       implementor expected to be imported", i.e. "blow up if you use something
       undocumented or private".  (Okay, this is at least two extra tickets,
       and I should probably be the one to file them, so consider this a "note
       to self" more than a feedback comment.  If it influences you to do
       anything, then great, but feel free to disregard it.)
    7. It's a bit weird that `iterAttributes` gives you `PythonAttribute`s, but
       `iterExportNames` only gives you `str`s.  I think for consistency it
       would be best to always return `PythonAttribute`s in the public API.
       You can easily get from there to names if you want, by using
       .name.split(".")[-1].  (OK, there should probably be something like
       'basename' on `PythonAttribute`, come to think of it.)  In light of
       this, you can of course amend my feedback from point 3 to be
       `exportedAttributes` and `importedAttributes`.  (Ooh!  Even better,
       `PythonAttribute` could have an `originalDefinition` attribute, which
       could point at the place where it was imported from!  But I digress,
       that is once again getting into a whole other complicated feature.)
    8. Minor nit: what is your fill-column set to?  All the new docstrings are
       wrapped either too short or too long.
    9. Also possibly for another ticket, but it would be nice if there were a
       "just do your best" mode for ignoring things like '`import *`' and
       `__all__` redefinition.  It's still possible that this code might be
       getting the wrong answer (for example: `__all__` defined statically, but
       then manipulated inside a function, or updated by another module) so a
       best guess wouldn't be all that different.  I suspect that most of the
       uses that this code would be put to would benefit from a `UserWarning`
       (or something like it - maybe a list of warning objects in the
       `PythonModule` itself) more than an exception.

If you can deal with the relatively small number of implementation / naming issues here, I'll commit to doing another quick review round soon afterward.  File as many of the additional tickets as you deem interesting.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [30277]) re [#4531](https://github.com/twisted/twisted/issues/4531) address first point of review feedback and clarify lookForSource.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [30279]) re [#4531](https://github.com/twisted/twisted/issues/4531) iter...Names() -> ...edNames() as per review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [30282]) re [#4531](https://github.com/twisted/twisted/issues/4531) per code review, maybe we shouldn't have a ticket for supporting 'import *', since I'm not sure it's a good idea, but if someone is considering adding the feature, it should be clearer that it's possible

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [39020]) Branching to 'static-modules-4531-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve removed owner|
|-|-|

OK, I merged forward the branch and fixed a couple of minor things. The 2.6 requirement makes the branch much cleaner as we can rely on having the ast module now. It seems most review points have been handled, at least there were several API renaming, and test coverage seems good, so back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_16"></a>|@wallrj commented|
|-|-|

## Code Review

Notes:
 * Merges cleanly
 * Forced build. Results look good.:
   * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/static-modules-4531-2

Points:
 1. [branches/static-modules-4531-2/twisted/test/test_modules.py](branches/static-modules-4531-2/twisted/test/test_modules.py)
    1. Test module names a.py, b.py make the tests difficult to
       follow. Why not use the same descriptive names that are used
       for the module content eg module_with_exported_imports
    2. I note that many of the tests use multiple assertions and
       assertions within loops which is now frowned upon
       see ticket:3926#comment:38, ticket:4392#comment:17,
       ticket:5668#comment:17.

 2. [branches/static-modules-4531-2/twisted/python/modules.py](branches/static-modules-4531-2/twisted/python/modules.py)
    1. PythonModule._maybeLoadFinder: missing coverage
       ```
       except TypeError
           raise ValueError
       ```
    2. PythonPath.__getitem__: missing coverage
       ```
            if path is None:
                path = moduleObject.__file__
       ```
    3. PythonModule.exported:
       1. I notice that this may raise SyntaxError first time round
          then succeed the next time. eg
          ```
          [richard@zorin static-modules-4531-2]$ python
          Python 2.7.5 (default, Jul  8 2013, 09:48:59)
          [GCC 4.8.1 20130603 (Red Hat 4.8.1-1)] on linux2
          Type "help", "copyright", "credits" or "license" for more information.
          >>> from twisted.python.modules import getModule
          >>> m = getModule('twisted.application.internet')
          >>> m.exported()
          Traceback (most recent call last):
            File "<stdin>", line 1, in <module>
            File "twisted/python/modules.py", line 532, in exported
              self._maybeLoadFinder()
            File "twisted/python/modules.py", line 476, in _maybeLoadFinder
              self._finder.visit(tree)
            File "/usr/lib64/python2.7/ast.py", line 241, in visit
              return visitor(node)
            File "twisted/python/modules.py", line 335, in visit_Module
              collectNames(stmt.targets[0], stmt.value)
            File "twisted/python/modules.py", line 301, in collectNames
              collectSingleName(value, target.id)
            File "twisted/python/modules.py", line 316, in collectSingleName
              checkAll(value)
            File "twisted/python/modules.py", line 326, in checkAll
              raise SyntaxError("__all__ must only contain literal"
          SyntaxError: __all__ must only contain literal Python identifier strings

          >>> m.exported()
          [PythonAttribute<'CooperatorService'>, PythonAttribute<'StreamServerEndpointService'>, PythonAttribute<'TimerService'>]
          ```
       2. It seems inconsistent that iterAttributes and exported
          generate different module names.
          ```
          [richard@zorin static-modules-4531-2]$ cat foo.py
          class Foo(object):
              __bar = 'baz'

          __Bar = 'baz'

          ```

          ```
          In [1]: from twisted.python.modules import getModule
          In [2]: m = getModule('foo')
          ...
          In [4]: m.isLoaded()
          Out[4]: False
          ```

          ```
          In [5]: list(m.iterAttributes())
          Out[5]: [PythonAttribute<'foo.__Bar'>, PythonAttribute<'foo.Foo'>]

          In [6]: m.exported()
          Out[6]: [PythonAttribute<'Foo'>]
          ```

       3. It also seems inconsistent that importedNames returns a
          generator while exported returns a list
          ```
          In [6]: m.exported()
          Out[6]: [PythonAttribute<'Foo'>]
          ```

          ```
          In [7]: m.importedNames()
          Out[7]: <generator object importedNames at 0x267d2d0>
          ```

       4. static analysis doesn't work for things like
          twisted.application.internet and
          twisted.internet.test.reactormixins which do
          globals().update(). Which would be an issue if it was to be
          used by pydoctor or trial.

       5. _isPythonIdentifier
          1. returns True for ''. Is that right? And
             I can't see a test for that string.

          2. I wonder if there's anything in the compiler module that
             we could re-use here.
             ```
             In [38]: compiler.parse('x = y')
             Out[38]: Module(None, Stmt([Assign([AssName('x', 'OP_ASSIGN')], Name('y'))]))

             In [39]: compiler.parse('123xyz = y')
               File "<string>", line 1
                 123xyz = y
                      ^
             SyntaxError: invalid syntax
             ```

 3. Generally I found it a bit confusing when the API gives different
    results depending on whether the object has been loaded.

    1. Would it be worth considering having separate classes and
       functions for static and non-static introspection? ie
       getModuleStatic, PythonModuleStatic, PythonAttributeStatic

    2. At least add some docstrings explaining (with examples)  the
       difference between static and loaded behaviour.

 4. Missing documentation. It would be very nice to have some lore
    documentation for twisted.python.modules. You could add some here
    or raise a ticket for that.

Overall, this looks like a really useful addition which I'm sure will
be useful in pydoctor, twistedchecker, trial etc.

I see that exarkun and glyph have both reviewed this already and
seemed reasonably happy with the behaviour... and glyph has addressed
a number of his own code review points.

So maybe this is ok to merge after addressing or answering the points
above. I'll leave it up to you.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

Replying to [rwall](#note_16):

Thanks for the great review! I didn't handle it as well as I'd like, but here it is:

> Points:
>  1. [branches/static-modules-4531-2/twisted/test/test_modules.py](branches/static-modules-4531-2/twisted/test/test_modules.py)
>     1. Test module names a.py, b.py make the tests difficult to
>        follow. Why not use the same descriptive names that are used
>        for the module content eg module_with_exported_imports

Fixed.

>     2. I note that many of the tests use multiple assertions and
>        assertions within loops which is now frowned upon
>        see ticket:3926#comment:38, ticket:4392#comment:17,
>        ticket:5668#comment:17.

This is a pretty big refactoring, so I left it out. You know I'd rather have a test per assert, but almost all of them use it, so it would basically mean restarting the branch from scratch.


>  2. [branches/static-modules-4531-2/twisted/python/modules.py](branches/static-modules-4531-2/twisted/python/modules.py)
>     1. PythonModule._maybeLoadFinder: missing coverage
>        ```
>        except TypeError
>            raise ValueError
>        ```
>     2. PythonPath.__getitem__: missing coverage
>        ```
>             if path is None:
>                 path = moduleObject.__file__
>        ```

Fixed by removing the code.

>     3. PythonModule.exported:
>        1. I notice that this may raise SyntaxError first time round
>           then succeed the next time. eg
>           ```
>           [richard@zorin static-modules-4531-2]$ python
>           Python 2.7.5 (default, Jul  8 2013, 09:48:59)
>           [GCC 4.8.1 20130603 (Red Hat 4.8.1-1)] on linux2
>           Type "help", "copyright", "credits" or "license" for more information.
>           >>> from twisted.python.modules import getModule
>           >>> m = getModule('twisted.application.internet')
>           >>> m.exported()
>           Traceback (most recent call last):
>             File "<stdin>", line 1, in <module>
>             File "twisted/python/modules.py", line 532, in exported
>               self._maybeLoadFinder()
>             File "twisted/python/modules.py", line 476, in _maybeLoadFinder
>               self._finder.visit(tree)
>             File "/usr/lib64/python2.7/ast.py", line 241, in visit
>               return visitor(node)
>             File "twisted/python/modules.py", line 335, in visit_Module
>               collectNames(stmt.targets[0], stmt.value)
>             File "twisted/python/modules.py", line 301, in collectNames
>               collectSingleName(value, target.id)
>             File "twisted/python/modules.py", line 316, in collectSingleName
>               checkAll(value)
>             File "twisted/python/modules.py", line 326, in checkAll
>               raise SyntaxError("__all__ must only contain literal"
>           SyntaxError: __all__ must only contain literal Python identifier strings
> 
>           >>> m.exported()
>           [PythonAttribute<'CooperatorService'>, PythonAttribute<'StreamServerEndpointService'>, PythonAttribute<'TimerService'>]
>           ```

Fixed.

>        2. It seems inconsistent that iterAttributes and exported
>           generate different module names.
>           ```
>           [richard@zorin static-modules-4531-2]$ cat foo.py
>           class Foo(object):
>               __bar = 'baz'
> 
>           __Bar = 'baz'
> 
>           ```
> 
>           ```
>           In [1]: from twisted.python.modules import getModule
>           In [2]: m = getModule('foo')
>           ...
>           In [4]: m.isLoaded()
>           Out[4]: False
>           ```
> 
>           ```
>           In [5]: list(m.iterAttributes())
>           Out[5]: [PythonAttribute<'foo.__Bar'>, PythonAttribute<'foo.Foo'>]
> 
>           In [6]: m.exported()
>           Out[6]: [PythonAttribute<'Foo'>]
>           ```

Fixed.

>        3. It also seems inconsistent that importedNames returns a
>           generator while exported returns a list
>           ```
>           In [6]: m.exported()
>           Out[6]: [PythonAttribute<'Foo'>]
>           ```
> 
>           ```
>           In [7]: m.importedNames()
>           Out[7]: <generator object importedNames at 0x267d2d0>
>           ```

I don't think it matters too much.

>        4. static analysis doesn't work for things like
>           twisted.application.internet and
>           twisted.internet.test.reactormixins which do
>           globals().update(). Which would be an issue if it was to be
>           used by pydoctor or trial.

I suspect we can fix twisted.application.internet by making it a static list. I'll open a bug once that one is merged, so that we can decide what to do.

>        5. _isPythonIdentifier
>           1. returns True for ''. Is that right? And
>              I can't see a test for that string.

Fixed.

>           2. I wonder if there's anything in the compiler module that
>              we could re-use here.
>              ```
>              In [38]: compiler.parse('x = y')
>              Out[38]: Module(None, Stmt([Assign([AssName('x', 'OP_ASSIGN')], Name('y'))]))
> 
>              In [39]: compiler.parse('123xyz = y')
>                File "<string>", line 1
>                  123xyz = y
>                       ^
>              SyntaxError: invalid syntax
>              ```

I kept the simple, albeit slightly broken, behavior for now.

>  3. Generally I found it a bit confusing when the API gives different
>     results depending on whether the object has been loaded.
> 
>     1. Would it be worth considering having separate classes and
>        functions for static and non-static introspection? ie
>        getModuleStatic, PythonModuleStatic, PythonAttributeStatic

Unfortunately the API has somewhat been designed this way, not new in this branch.

>     2. At least add some docstrings explaining (with examples)  the
>        difference between static and loaded behaviour.
> 
>  4. Missing documentation. It would be very nice to have some lore
>     documentation for twisted.python.modules. You could add some here
>     or raise a ticket for that.

Agreed. I don't think we can document all of it here though. Would something in the docstring mentioned exported() and importedNames() be enough for now?


Back for another round.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj set owner to @therve<br>@wallrj set status to `new`|
|-|-|

Thanks therve.

All the changes since the last review look good and the behaviour is
now more consistent.

I said it in the last review, but I don't really like that
"importedNames returns a generator while exported returns a list"
but it doesn't really matter I suppose.

Notes:
 * Merges cleanly
 * Build results are all green (apart from some spurious warnings)

Some remaining minor points:
 1. Add a short paragraph to the twisted.python.modules docstring
    summarising which types of object can be analysed statically.
 2. Create a ticket "fix twisted.application.internet by making it a
    static list" (comment:17) -- maybe the ticket title should be
    "replace dynamically assigned module attributes to allow static
    analysis with t.p.modules"
 3. Create a ticket for Glyph's suggestion "unambiguously request the scan-the-source
    behavior, because sometimes it differs from the
    inspect-the-loaded-module behavior"  (comment:9) -- I like that
    idea.

Please merge after addressing (or answering) those numbered points.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj commented|
|-|-|

Oh yeah, and add a .feature news file. Maybe "twisted.python.modules now supports static analysis of module attributes"

**Comment by tristanlatr** [comment]
Should this issue be close ?

**Comment by adiroiban** [comment]
I guess that at this point, is best to have `code analysis` functionality outside of Twisted.

twisted/twisted should be focuses on the networking part

I guess that 14 years ago the stdlib ast and other tools were not that good, so there was this PR to add code to twisted.

I am closing this now, but the branch is not removed.

We can reopen this if people think that Twisted should have code analysis helpers.

**Comment by glyph** [comment]
twisted needs `twisted.python.modules` for its plugin system. If it would make sense to broaden the API so that we have a consistent abstraction, perhaps we could move this out to the new library that @tristanlatr is starting and have a common dependency between twisted and pydoctor.
